Dynamically update the brand colors

Ridhwana opened the pull request:
Dynamically update the brand colors  on the creator settings page when a brand color is selected

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

When a user updates the brand color, we update the brand on the page to reflect that color.

![color](https://user-images.githubusercontent.com/2786819/142413973-1cf4f88b-a7ab-4b29-971c-3146dc5893a6.gif)

This includes updating: 
- The Forem background image 
- The help button 
- The finish button
- the focus on the radio button 
- the accents on the text fields

## Related Tickets & Documents
Closes https://github.com/forem/forem/issues/14646

## QA Instructions, Screenshots, Recordings

#### Please note that you will need to essentially clear out your database to take this feature for a spin. 🏎️  

To do so, you can either drop your entire database via `rails:db drop` 

_**or**_

You can open a rails console via `rails c` and do the following:
 ```
# Enable the feature flag
FeatureFlag.enable(:creator_onboarding)

# Delete all users 
User.destroy_all

# Set `waiting_on_first_user` to be true
Settings::General.waiting_on_first_user = true

# If you've logged in before set the mascot id to nil
 Settings::General.mascot_user_id = nil
```

Now, run your server and navigate to `http://localhost:3000/` and complete the "Create Account" form. After completing the "Create Account" form, you should be brought to the new Creator Setup page, `${baseUrl}/admin/creator_settings/new?referrer=${baseUrl}`.

Once on the Creator Setup page, you will want to observe the changes outlined below:

When you click on the color picker and choose a new color, and then close it OR you enter a hex color in the text field you should see the color being updated on the following elements:
- The Forem background image 
- The help button 
- The finish button
- the focus on the radio button 
- the accents on the text fields

### UI accessibility concerns?

_If your PR includes UI changes, please replace this line with details on how
accessibility is impacted and tested. For more info, check out the
[Forem Accessibility Docs](https://developers.forem.com/frontend/accessibility)._

## Added/updated tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://developers.forem.com) or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [x] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] Are there any post deployment tasks we need to perform?

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](https://media.giphy.com/media/nzcb2ZQ2L48Za/giphy.gif)

Ridhwana commented on app/assets/images/forem-background.svg:
```diff
@@ -1,3 +1,3 @@
-<svg width="475" height="523" viewBox="0 0 475 523" fill="none" xmlns="http://www.w3.org/2000/svg">
```
We remove the fill so that we can use the brand color accent color settings

Ridhwana commented on app/javascript/admin/controllers/creator_settings_controller.js:
```diff
@@ -67,4 +67,53 @@ export class CreatorSettingsController extends Controller {
+    // We need to recalculate this in javascript as it's currently being calculated in ruby
+    document.documentElement.style.setProperty(
+      '--accent-brand-darker',
+      this.updatedBrightness(color, 0.85),
```
For the hover effect on buttons we need to calculate a new --accent-brand-darker which decreases the brightness by 85%.

Ridhwana replied:
Done ✅

Ridhwana commented on app/views/admin/creator_settings/_form.html.erb:
```diff
@@ -52,11 +54,11 @@
     <div>
       <div class="mb-2">
-        <%= radio_button_tag :invite_only_mode, "0", class: "crayons-field crayons-field--radio", required: true %>
+        <%= radio_button_tag :invite_only_mode, "0", false, class: "crayons-radio" %>
```
The crayons-radio styles allow the radio buttons to receive the brand color styling. In addition, the required fields weren't being applied here.

Ridhwana replied:
We added it to honor the positional params for a radio_tag_button:
```
radio_button_tag(name, value, checked, options) 
```

Prior to this change, the class `crayons-field crayons-field--radio` wasn't being applied at all because we were not honouring the position of the arg. However, the radio button was still not being checked because the default for the checked arg is set to false anyways.

Hence, `false` is to indicate that we don't want this radio button to be checked.

Ridhwana commented on app/views/users/show.html.erb:
```diff
@@ -33,7 +33,7 @@
 
           <div class="profile-header__actions">
               <button id="user-follow-butt" class="crayons-btn whitespace-nowrap follow-action-button follow-user" data-info='{"id":<%= @user.id %>,"className":"<%= @user.class.name %>", "name": "<%= @user.name %>"}'><%= t("views.users.follow") %></button>
-	      <div class="profile-dropdown ml-2 s:relative hidden" data-username="<%= @user.username %>">
```
This is unrelated, it was just a rubucop change.

Ridhwana commented on app/views/devise/confirmations/new.html.erb:
```diff
@@ -22,7 +22,7 @@
     </div>
 
-    <%= inline_svg_tag("forem-background.svg", aria: true, title: "forem background", class: "absolute bottom-0 right-0 hidden m:block") %>
+    <%= inline_svg_tag("forem-background.svg", aria: true, title: "forem background", class: "forem-background absolute bottom-0 right-0 hidden m:block") %>
```
We add a `forem-background` class so that we can set the `brand-accent` color for the background across all the views using it.

ludwiczakpawel commented on app/javascript/admin/controllers/creator_settings_controller.js:
```diff
@@ -67,4 +67,53 @@ export class CreatorSettingsController extends Controller {
+    );
+  }
+
+  updatedBrightness(color, amount = 1) {
+    const rgbObj = this.hexToRgb(color);
+    Object.keys(rgbObj).forEach((key) => {
+      rgbObj[key] = Math.round(rgbObj[key] * amount);
+    });
+
+    return this.rgbToHex(rgbObj['r'], rgbObj['g'], rgbObj['b']);
+  }
+
+  hexToRgb(hex) {
+    const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
+    return result
+      ? {
+          r: parseInt(result[1], 16),
+          g: parseInt(result[2], 16),
+          b: parseInt(result[3], 16),
+        }
+      : null;
+  }
+
+  rgbToHex(r, g, b) {
+    return `#${this.rgbParameterToHex(r)}${this.rgbParameterToHex(
+      g,
+    )}${this.rgbParameterToHex(b)}`;
+  }
+
+  rgbParameterToHex(param) {
+    const hex = param.toString(16);
+    return hex.length == 1 ? `0${hex}` : hex;
+  }
```
Any chance we could make these functions somehow global/shared? I'm pretty sure we would benefit from these elsewhere, like for tags coloring or for future themes customization for admins, etc.. pretty_please.gif

Ridhwana replied:
I did think of it however I didn't do that because I didn't want to pre-emptively generalize that we would use that elsewhere and thought that it could be pulled out when we needed it.  

I'm happy to do it though if we think that there is a need for it. 
@nickytonline do you have a suggestion on where in the codebase a generalized function like this would live?

ludwiczakpawel replied:
No **strong** opinion here... I personally think we should do it, because it's possible no one will actually know/remember that we have this function existing here when we get to implement that elsewhere (places I've mentioned earlier). It's more likely to be discoverable when placed in shared/global place.  But that's just my $0.02 - as I said - it's not the hill to die on for me :D.

Ridhwana replied:
That rationale makes sense to me :) You're right about people forgetting it exists 😁 I don't think this should hold the PR from being approved but I will simultaneously move this out into a global function. Thanks for bringing it up @ludwiczakpawel.

aitchiss replied:
I think in a case like this it's quite good to split it out into a utility purely for being able to unit test it in isolation  (which I think would be really useful for these helpers)

Ridhwana replied:
@aitchiss sounds good, and a unit test for this will be useful as well. 
 
Regarding the placement of the utility file in the codebase, does this seem like a good [folder](https://github.com/forem/forem/tree/main/app/javascript/admin) to put it in alongside messageUtilities.js

aitchiss replied:
Yeah I think that probably makes most sense at the moment, since we're not anticipating (yet) these utils being used outside of admin areas 👍

nickytonline replied:
We have a utilities folder in the frontend. Consider `app/javascript/utilities/colors.js`

Ridhwana replied:
Done ✅

aitchiss commented on app/javascript/admin/controllers/creator_settings_controller.js:
```diff
@@ -67,4 +67,53 @@ export class CreatorSettingsController extends Controller {
+  }
+
+  hexToRgb(hex) {
+    const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
```
(not blocking) maybe it could be useful to add a comment about what's happening in this regex

Ridhwana replied:
I added some descriptions for the functions in the utility classes to help with this ✅

nickytonline commented on cypress/integration/creatorOnboardingFlows/creatorSettings.spec.js:
```diff
@@ -59,6 +59,37 @@ describe('Creator Settings Page', () => {
+
+    cy.findAllByRole('radio', { name: /members only/i })
+      .check()
+      .should('have.css', 'background-color', rgbColor)
```
As discussed yesterday, we typically haven't tested for color in E2E tests, but in this particular case, I think it makes sense.

msarit replied:
For some reason, I do not see this spec in Cypress 😟

juliannatetreault replied:
Do you have the latest `main`, @msarit? I just double-checked that I still see the spec file and verified that I do. 😕

msarit replied:
ok i will update my local version of Ridhwana's branch with `main`

nickytonline commented on cypress/integration/creatorOnboardingFlows/creatorSettings.spec.js:
```diff
@@ -59,6 +59,37 @@ describe('Creator Settings Page', () => {
+      .focus()
+      .should('have.css', 'border-color', rgbColor);
+
+    cy.get('.admin-help-button').should(
```
Refactor (blocking) use `cy.findByRole instead as it's more reliable than a CSS class name which may change in the future.

```suggestion
    cy.findByRole('link', { name: /community name/i }).should(
```

nickytonline reviewed (changes requested):
This is working great. One small change request.

nickytonline commented on app/javascript/admin/controllers/creator_settings_controller.js:
```diff
@@ -67,4 +67,53 @@ export class CreatorSettingsController extends Controller {
+   *
+   * @param {Event} event
+   */
+  updateBranding(event) {
```
Refactor (blocking): The `pattern` attribute on the element should be used in the code that changes the colour on the screen.

If the pattern doesn't match, return early.

e.g. in creator_settings_controller

```javascript
  updateBranding(event) {
    const { value: color } = event.target;

    if (!new RegExp(event.target.getAttribute('pattern')).test(color)) {
      return;
    }
    
    ...
 }
```

Ridhwana replied:
Done ✅

nickytonline replied:
Nitpick (non-blocking): Consider `updateBrandingColor` instead.

Ridhwana replied:
@nickytonline the name of this action needs to change in my next PR where this action will serve to update branding and  check color contrast, I'll update it there. Thanks!

Ridhwana commented on app/javascript/utilities/color/accentCalculator.js:
```diff
@@ -0,0 +1,58 @@
+ * Based on the ruby implementation
+ * https://github.com/forem/forem/blob/main/app/services/color/compare_hex.rb
+ */
+export function brightness(color, amount = 1) {
```
I added a utility class that calculates the brightness as discussed. The reason I put in a folder called `color` is because there will be more color utilities like validating the color contrast (which I'm working on in parallel)  so grouping them will be useful.

Ridhwana commented:
@aitchiss , @nickytonline , @ludwiczakpawel I've addressed all the comments and this is ready to be reviewed again. One of the more notable changes I've made is adding the `brightness` change to a utility file and writing a unit test for it. I've also included the suggestions and updates for other comments.

juliannatetreault reviewed (approved):
This looks great! 😍  I tested this out locally and it worked like a charm! ✨

nickytonline reviewed (approved):
Looks great @ridhwana! 🚀
